pkg/router/controller/route_secret_manager_test.go

```diff
@@ -0,0 +1,1139 @@
+						},
+					},
+				},
+			},
+			secretManager: fake.SecretManager{
+				Secret: fakeSecret("sandbox", "tls-secret", corev1.SecretTypeBasicAuth, map[string][]byte{}),
+			},
+			eventType: watch.Added,
+			allow:     true,
+			expectedRoute: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{
+					TLS: &routev1.TLSConfig{
+						ExternalCertificate: &routev1.LocalObjectReference{
+							Name: "tls-secret",
+						},
+					},
+				},
+			},
+			expectedEventType: watch.Deleted,
+			expectedRejections: map[string]string{
+				"sandbox-route-test": "ExternalCertificateValidationFailed",
+			},
+			expectedError: true,
+		},
+		{
+			name: "route added with externalCertificate allowed and correct secret",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{
+					TLS: &routev1.TLSConfig{
+						ExternalCertificate: &routev1.LocalObjectReference{
+							Name: "tls-secret",
+						},
+					},
+				},
+			},
+			secretManager: fake.SecretManager{
+				Secret: fakeSecret("sandbox", "tls-secret", corev1.SecretTypeTLS, map[string][]byte{
+					"tls.crt": []byte("my-crt"),
+					"tls.key": []byte("my-key"),
+				}),
+			},
+			eventType: watch.Added,
+			allow:     true,
+			expectedRoute: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{
+					TLS: &routev1.TLSConfig{
+						ExternalCertificate: &routev1.LocalObjectReference{
+							Name: "tls-secret",
+						},
+						Certificate: "my-crt",
+						Key:         "my-key",
+					},
+				},
+			},
+			expectedEventType:  watch.Added,
+			expectedRejections: map[string]string{},
+			expectedError:      false,
+		},
+		{
+			name: "route added with externalCertificate allowed and correct secret but got error from secretManager",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{
+					TLS: &routev1.TLSConfig{
+						ExternalCertificate: &routev1.LocalObjectReference{
+							Name: "tls-secret",
+						},
+					},
+				},
+			},
+			secretManager: fake.SecretManager{
+				Secret: fakeSecret("sandbox", "tls-secret", corev1.SecretTypeTLS, map[string][]byte{
+					"tls.crt": []byte("my-crt"),
+					"tls.key": []byte("my-key"),
+				}),
+				Err: fmt.Errorf("something"),
+			},
+			eventType:          watch.Added,
+			allow:              true,
+			expectedRejections: map[string]string{},
+			expectedError:      true,
+		},
+		{
+			name: "route updated: old route without externalCertificate, new route without externalCertificate",
+			route: &routev1.Route{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "route-test",
+					Namespace: "sandbox",
+				},
+				Spec: routev1.RouteSpec{},
+			},
```

Comment:
This is the new route I suppose. Where is the definition of the old one? How can I understand the scenario tests what it states in the name?

Reply:
Yes, this is the new route. 

The reason behind not having the old route spec is that we never pass the old route in the plugin chain while calling `HandleRoute()`. It’s only the new route that propagates in the plugin chain. So, the definition for the old route will be of no use in the test case scenarios also. 

https://github.com/openshift/router/blob/56ab14f257e123317c00ce7ba5b4ecae7b2e0102/pkg/router/interfaces.go#L13-L20

The only way to know whether the route was registered earlier or not, is to rely on `SecretManager`’s `IsRouteRegistered()` function. And so, we have `IsRegistered: true / false`  in the `fake.SecretManager` definition, to indicate that in the test cases.

So, if the old route was having `externalCertificate` , we will define 

```go
			secretManager: fake.SecretManager{
				IsRegistered: true,
			},
```

Similarly, when route is deleted.